crates/bevy_ecs/src/component.rs

```diff
@@ -401,6 +403,41 @@ pub trait Component: Send + Sync + 'static {
     }
 }
 
+// NOTE: This should maybe be private, but it is currently public so that `bevy_ecs_macros` can use it.
+// This exists as a standalone function instead of being inlined into the component derive macro so as
+// to reduce the amount of generated code.
+#[doc(hidden)]
+pub fn enforce_no_required_components_recursion(
+    components: &Components,
+    recursion_check_stack: &[ComponentId],
+) {
+    if let Some((&requiree, check)) = recursion_check_stack.split_last() {
+        if let Some(direct_recursion) = check
+            .iter()
+            .rev()
+            .enumerate()
+            .find_map(|(index, &id)| (id == requiree).then_some(index == 0))
```

Comment:
Just a nit, but this could be a little shorter using `rposition`

```suggestion
            .rposition(|&id| id == requiree)
            .map(|index| index == 0)
```